### PR TITLE
ARROW-6839: [Java] Add APIs to read and write "custom_metadata" field of IPC file footer

### PR DESCRIPTION
**Pull request by tianchen92**
Related to [ARROW-6839](https://issues.apache.org/jira/browse/ARROW-6839).

This is the java implementation of format change in ARROW-6836.

### CODE REVIEW

**Comment by tianchen92** [comment]
@wesm @jacques-n @emkornfield Please help take a look when you have time, thanks!

**Comment by wesm** [comment]
cc @lidavidm also

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java:770]: is dictionary encoding really necessary for this test?  Can you use the VectorPopulator is any data is actually needed?

Comment (tianchen92) [reply java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java]: >is dictionary encoding really necessary for this test?

Not necessary, removed, thanks.

>Can you use the VectorPopulator is any data is actually needed?

Hmm, always forget this, fixed.

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/ArrowFileReader.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/ipc/ArrowFileReader.java:123]: is this consistent with other methods, I think returning an empty map might be better?

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowFooter.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowFooter.java:143]: i'm suprised this code doesn't exist else where for converted meta between string,string map

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java:763]: are any values needed?  If so i imagine just one is needed?

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java:775]: does a batch even need to be written?

**Review by emkornfield — CHANGES_REQUESTED** [review]

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/ArrowFileReader.java
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/ipc/ArrowFileReader.java]: ok, fixed.

File: java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowFooter.java
Comment (tianchen92) [reply java/vector/src/main/java/org/apache/arrow/vector/ipc/message/ArrowFooter.java]: Actually similar logic exist in Schema for writing metadata, I extracted a static write method in FBSerializables for reuse purpose.

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java
Comment (tianchen92) [reply java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java]: ok, no special reason, made it look like a normal file data. Removed now.

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java
Comment (tianchen92) [reply java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowReaderWriter.java]: also removed, thanks

**Review by rymurr — APPROVED** [review]
+1

**Comment by tianchen92** [comment]
Thanks @rymurr for the review.
@emkornfield Do you have other comments? otherwise I'll merge this in several days later.

**Comment by emkornfield** [comment]
No please go ahead and merge.

**Comment by emkornfield** [comment]
@tianchen92 did you use the merge tool to merge this (if you haven't please read https://cwiki.apache.org/confluence/display/ARROW/Guide+for+Committers+and+Project+Maintainers)

**Comment by tianchen92** [comment]
> @tianchen92 did you use the merge tool to merge this (if you haven't please read https://cwiki.apache.org/confluence/display/ARROW/Guide+for+Committers+and+Project+Maintainers)

Hmm, I directly merged this by click "squash and merge" button, should we revert this?

**Comment by emkornfield** [comment]
Not familiar enough with the implications @kszucs @kou what are the correct steps here?

**Comment by wesm** [comment]
The commit looks OK but @tianchen92 please use the merge tool not the GitHub UI to merge patches.

**Comment by tianchen92** [comment]
> The commit looks OK but @tianchen92 please use the merge tool not the GitHub UI to merge patches.

I see, sorry for the incorrect merge operation :)

**Pull request merged**
